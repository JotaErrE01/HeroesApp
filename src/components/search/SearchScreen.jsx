import React from 'react'
import { useLocation } from 'react-router';
import useFrom from '../../hooks/useForm';
import HeroCard from '../heroes/HeroCard';
import queryString from 'query-string';
import getHeroByName from '../../selectors/getHeroByName';

const SearchScreen = ({ history }) => {

    const { search } = useLocation();

    const { q = '' } = queryString.parse(search);
    
    const [ { hero }, handleInputChange ] = useFrom({
        hero: q
    });

    const heroesFiltered = getHeroByName(q); // pasar hero si deseas buscar mientras escribes

    const handleSearch = e => {
        e.preventDefault();
        history.push(`?q=${hero}`);
    }

    return (
        <div>
            <h1>Search Screen</h1>
            <hr />

            <div className="row">
                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr />

                    <form
                        onSubmit={handleSearch}
                    >
                        <input 
                            type="text" 
                            placeholder="Find your hero"
                            className="form-control"
                            autoComplete="off"
                            name="hero"
                            value={hero}
                            onChange={handleInputChange}
                        />

                        <button
                            type="submit"
                            className="btn btn-block m-1 btn-outline-primary"
                        >Search</button>
                    </form>
                </div>

                <div className="col-7">
                    <h4>Results</h4>
                    <hr />

                    {
                        q.length === 0 &&

                        <div className="alert alert-info">
                            Search a Hero
                        </div>
                    }

                    {
                        heroesFiltered.length === 0 && 
                        
                        q.length > 0 &&

                        <div className="alert alert-danger">
                            Threre is not hero with { q }
                        </div>
                    }

                    {
                        heroesFiltered.map(hero => (
                            <HeroCard 
                                key={hero.id}
                                {...hero}
                            />
                        ))
                    }
                </div>
            </div>
        </div>
    )
}

export default SearchScreen;