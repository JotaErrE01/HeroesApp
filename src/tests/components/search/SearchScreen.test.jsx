import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter, Route } from 'react-router';
import SearchScreen from '../../../components/search/SearchScreen';

describe('Pruebas en el <SearchScreen />', () => {
    
    test('Debe de hacer match con el snapshot', () => {
        
        const { container } = render(
            <MemoryRouter>
                <SearchScreen />
            </MemoryRouter>
        );

        expect( container ).toMatchSnapshot();

    });

    test('Debe de mostrar a batman y el valor del query String', () => {
        
        const { container } = render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route exact path='/search' component={SearchScreen} />
            </MemoryRouter>
        )
        expect( screen.getByPlaceholderText('Find your hero').value ).toBe('batman');
        expect( container.querySelector('.card-tittle').textContent.trim().toLowerCase() ).toBe('batman');

    });

    test('Debe de mostrar un error si no se encuentra el hero', () => {
        
        const { container } = render(
            <MemoryRouter initialEntries={['/search?q=batmanjsdf;']}>
                <Route exact path='/search' component={SearchScreen} />
            </MemoryRouter>
        )

        expect( container.querySelector('.alert-danger') ).toBeInTheDocument();

    })

    test('Debe de llamar al push del history', () => {

        const history = {
            push: jest.fn()
        }
        
        const { container } = render(
            <MemoryRouter initialEntries={['/search']}>
                <Route 
                    exact 
                    path='/search' 
                    component={ props => <SearchScreen {...props} history={history}/> } 
                />
            </MemoryRouter>
        )

        userEvent.type(container.querySelector('.form-control'), 'batman');

        userEvent.click(container.querySelector('.btn-outline-primary'));

        expect( history.push ).toHaveBeenCalled();
        expect( history.push ).toHaveBeenCalledWith('?q=batman');

    })
    
})
