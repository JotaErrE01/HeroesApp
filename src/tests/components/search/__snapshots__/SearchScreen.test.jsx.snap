// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <SearchScreen /> Debe de hacer match con el snapshot 1`] = `
<div>
  <div>
    <h1>
      Search Screen
    </h1>
    <hr />
    <div
      class="row"
    >
      <div
        class="col-5"
      >
        <h4>
          Search Form
        </h4>
        <hr />
        <form>
          <input
            autocomplete="off"
            class="form-control"
            name="hero"
            placeholder="Find your hero"
            type="text"
            value=""
          />
          <button
            class="btn btn-block m-1 btn-outline-primary"
            type="submit"
          >
            Search
          </button>
        </form>
      </div>
      <div
        class="col-7"
      >
        <h4>
          Results
        </h4>
        <hr />
        <div
          class="alert alert-info"
        >
          Search a Hero
        </div>
      </div>
    </div>
  </div>
</div>
`;
